qiita_plugins/target_gene_type/target_gene_type/summary.py

```diff
@@ -0,0 +1,173 @@
+    -------
+    dict
+        The results of the job
+
+    Raises
+    ------
+    ValueError
+        - If there is any error gathering the information from the server
+        - If there artifact is 'Demultiplexed' but it doesn't have a demux file
+    """
+    # Step 1: gather file information from qiita using REST api
+    # 1a. getting the file paths
+    artifact_id = parameters['input_data']
+    qclient_url = "/qiita_db/artifacts/%s/filepaths/" % artifact_id
+    fps_info = qclient.get(qclient_url)
+    if not fps_info or not fps_info['success']:
+        error_msg = "Could not get artifact filepath information: %s"
+        if fps_info:
+            error_msg = error_msg % fps_info['error']
+        else:
+            error_msg = error_msg % "could not connect with the server"
+        raise ValueError(error_msg)
+    # 1.b get the artifact type_info
+    type_info = qclient.get("/qiita_db/artifacts/%s/type/" % artifact_id)
+    if not type_info or not type_info['success']:
+        error_msg = "Could not get artifact metadata information %s"
+        if type_info:
+            error_msg = error_msg % type_info['error']
+        else:
+            error_msg = error_msg % "could not connect with the server"
+        raise ValueError(error_msg)
+    artifact_type = type_info['type']
+
+    # we have 2 main cases: Demultiplexed and everything else, splitting on
+    # those
+    if artifact_type == 'Demultiplexed':
+        artifact_information = _summary_demultiplexed(
+            artifact_type, fps_info['filepaths'])
+        if artifact_information is None:
+            raise ValueError("We couldn't find a demux file in your artifact")
+    else:
+        artifact_information = _summary_not_demultiplexed(
+            artifact_type, fps_info['filepaths'])
+
+    of_fp = join(out_dir, "artifact_%d.html" % artifact_id)
+    of = open(of_fp, 'w')
+    of.write('\n'.join(artifact_information))
+
+    # Step 3: add the new file to the artifact using REST api
+    reply = qclient.patch(qclient_url, 'add', '/html_summary/',
+                          value=of_fp)
+
+    return reply if not return_html else (reply, artifact_information)
+
+
+def _summary_not_demultiplexed(artifact_type, filepaths):
+    # loop over each of the fps/fps_type pairs
+    artifact_information = []
+    for (fps, fps_type) in filepaths:
+        # Step 2: generate HTML summary
+        # md5, from http://stackoverflow.com/a/3431838
+        with open(fps, "rb") as f:
+            hash_md5 = md5()
+            for chunk in iter(lambda: f.read(4096), b""):
+                hash_md5.update(chunk)
+
+        # getting head of the files
+        header = []
+        if artifact_type not in FILEPATH_TYPE_TO_NOT_SHOW_HEAD:
+            # we need to encapsulate the full for loop because gzip will not
+            # raise an error until you try to read
+            try:
+                with gopen(fps, 'r') as fin:
+                    for i, line in enumerate(fin):
```

Comment:
Can you use the first example in [this link](http://stackoverflow.com/a/1767589/3746629) to read the first X lines of the file? Use also in line 122.